### DISCUSSION TITLE
The best way to use specific Linux kernel version

### DISCUSSION THREAD

**Discussion by tatry** [community/community#21317] [Actions] [answered]
I have a need to run tests under specific Linux kernel version in GitHub Actions. How to do this in the best way?

I want to avoid self-hosted runners due to maintenance & security reasons.

**Comment by AngusTripleH** [comment]
give up

**Comment by airtower-luna** [comment, marked as answer]
You can't change the kernel version for the hosted runner VMs, but you can run your own VMs inside them.

I'd probably use Qemu, or Qemu managed via libvirt. You can install both easily from the Ubuntu package repositories (via `apt-get`, etc.). The hard part will be to figure out what you want to boot, and how to integrate the tools to run the tests. Note that Qemu can boot a kernel image plus initramdisk directly, so you don't necessarily need one disk image per version. Then there's the question of how you want to send commands to the VM and get test results back, e.g. over the serial console, network, or shared folder (e.g. via p9fs).

What's best depends on what exactly your requirements are, but I hope this helps! :smiley_cat: :penguin:

**Reply by tatry** [reply]
Sounds like a good idea to me, I have to try it. Thank you!

**Reply by tatry** [reply]
For anybody interested in I have prepared a repository with demo: https://github.com/tatry/kernel-version-demo

**Reply by airtower-luna** [reply]
Nice, thank you for sharing your solution! :smile_cat:
